File: windows/Relay.App.Tests/SharedContracts.cs
using System.Text.Json;

namespace Relay.App.Tests;

/// <summary>
/// Locates the repo-level /shared directory from the test output directory,
/// so tests always run against the single source of truth.
/// </summary>
public static class SharedContracts
{
    public static readonly string Dir = Locate();

    public static JsonDocument Json(string name) =>
        JsonDocument.Parse(File.ReadAllText(Path.Combine(Dir, name)));

    private static string Locate()
    {
        var current = new DirectoryInfo(AppContext.BaseDirectory);
        while (current is not null)
        {
            var candidate = Path.Combine(current.FullName, "shared", "test-vectors.json");
            if (File.Exists(candidate)) return Path.Combine(current.FullName, "shared");
            current = current.Parent;
        }
        throw new InvalidOperationException(
            $"Could not locate /shared above {AppContext.BaseDirectory}");
    }
}

File: android/app/src/test/kotlin/io/relay/app/SharedContracts.kt
package io.relay.app

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import java.io.File

/**
 * Locates the repo-level /shared directory from the unit-test working
 * directory (module dir in Gradle), so tests always run against the single
 * source of truth instead of a copied fixture.
 */
object SharedContracts {
    val dir: File by lazy {
        var current: File? = File(System.getProperty("user.dir")).absoluteFile
        while (current != null) {
            val candidate = File(current, "shared/test-vectors.json")
            if (candidate.isFile) return@lazy File(current, "shared")
            current = current.parentFile
        }
        error("Could not locate /shared above ${System.getProperty("user.dir")}")
    }

    fun json(name: String): JsonElement =
        Json.parseToJsonElement(File(dir, name).readText(Charsets.UTF_8))
}
